File: frontend/src/utils/i18n.ts
/**
 * Clean, simple internationalization (i18n) utilities
 */

import { ref, computed } from 'vue'
import http from '@/utils/dynamic-http'

interface Translations {
  [key: string]: any
}

class FrontendI18n {
  private translations: Map<string, Translations> = new Map()
  private currentLanguage = ref('en')
  private version = ref(0)

  constructor() {
    // Initialize with empty English translations (will be populated by extensions)
    this.translations.set('en', {})
  }

  // Load translations from local extension files using dynamic scanning
  private async loadLocalTranslations(languageCode: string): Promise<Translations> {
    let allTranslations: Translations = {}

    try {
      // Use a comprehensive glob to capture all extension JSON files
      const allExtensionModules = import.meta.glob('../extensions/**/*.json', { eager: true }) as Record<string, any>

      // Find the language pack directory for this language code
      let languagePackBaseDir = ''

      for (const path in allExtensionModules) {
        if (path.includes('manifest.json')) {
          try {
            const manifest = allExtensionModules[path].default
            if (manifest.type === 'language' && manifest.language?.code === languageCode) {
              // Extract the base directory path (e.g., '../extensions/BulgarianLanguagePack_1.0.0')
              const pathParts = path.split('/')
              languagePackBaseDir = pathParts.slice(0, -1).join('/') // Remove manifest.json
              break
            }
          } catch (error) {
            // Continue
          }
        }
      }

      if (!languagePackBaseDir) {
        // No language pack found for this language, try loading from backend
        return await this.loadBackendTranslations(languageCode)
      }

      // Load all translation files from this language pack directory
      for (const path in allExtensionModules) {
        // Check if this file is in our language pack directory and not a manifest
        if (path.startsWith(languagePackBaseDir) && !path.includes('manifest.json')) {
          try {
            const translationData = allExtensionModules[path].default
            allTranslations = { ...allTranslations, ...translationData }
          } catch (error) {
            console.error(`Failed to load ${path}:`, error)
          }
        }
      }


      // If no translations loaded from local files, try backend
      if (Object.keys(allTranslations).length === 0) {
        allTranslations = await this.loadBackendTranslations(languageCode)
      }
    } catch (error) {
      console.error(`Failed to load local translations for ${languageCode}, trying backend:`, error)
      // Try loading from backend as fallback
      allTranslations = await this.loadBackendTranslations(languageCode)
    }

    return allTranslations
  }


  // Load translations from backend API
  private async loadBackendTranslations(languageCode: string): Promise<Translations> {
    try {
      // Import http dynamically to avoid circular dependency
      const { default: http } = await import('@/utils/dynamic-http')
      const response = await http.get(`/api/translations/${languageCode}`)
      const data = response.data
      return data.frontend || {}
    } catch (error) {
      console.error(`Failed to load backend translations for ${languageCode}:`, error)
      return {}
    }
  }

  // Deep merge function for nested objects
  private deepMerge(target: any, source: any): any {
    const result = { ...target }
    for (const key in source) {
      if (source[key] && typeof source[key] === 'object' && !Array.isArray(source[key])) {
        result[key] = this.deepMerge(result[key] || {}, source[key])
      } else {
        result[key] = source[key]
      }
    }
    return result
  }

  // Load translations for a language
  async loadTranslations(languageCode: string) {
    try {
      const localTranslations = await this.loadLocalTranslations(languageCode)
      const extensionTranslations = await this.loadExtensionTranslationsForLanguage(languageCode)
      const allTranslations = this.deepMerge(localTranslations, extensionTranslations)
      this.translations.set(languageCode, allTranslations)
      this.version.value++
    } catch (error) {
      console.error(`Failed to load translations for ${languageCode}:`, error)
    }
  }

  // Load extension translations for a specific language
  private async loadExtensionTranslationsForLanguage(languageCode: string): Promise<Translations> {
    let extensionTranslations: Translations = {}

    try {
      // Use glob to find all extension modules
      const allExtensionModules = import.meta.glob('../extensions/**/*.json', { eager: true }) as Record<string, any>

      // Find all extension directories that have locales
      const extensionDirs = new Set<string>()

      for (const path in allExtensionModules) {
        if (path.includes('manifest.json')) {
          try {
            const manifest = allExtensionModules[path].default
            if (manifest.locales && manifest.locales.supported?.includes(languageCode)) {
              // Extract the base directory path
              const pathParts = path.split('/')
              const baseDir = pathParts.slice(0, -1).join('/') // Remove manifest.json
              extensionDirs.add(baseDir)
            }
          } catch (error) {
            // Continue
          }
        }
      }

      // Load translations from each extension's locales directory
      for (const baseDir of extensionDirs) {
        const localesPath = `${baseDir}/locales/${languageCode}.json`
        if (localesPath in allExtensionModules) {
          try {
            const translationData = allExtensionModules[localesPath].default
            // Merge extension translations directly into root namespace for easier access
            extensionTranslations = this.deepMerge(extensionTranslations, translationData)
          } catch (error) {
            console.error(`Failed to load extension translations from ${localesPath}:`, error)
          }
        }
      }
    } catch (error) {
      console.error(`Failed to load extension translations for ${languageCode}:`, error)
    }

    return extensionTranslations
  }

  // Load extension translations for a specific extension (called when extension is loaded)
  async loadExtensionTranslationsForExtension(extensionName: string, languageCode: string) {
    try {
      // Use glob to find the extension's locale file
      const extensionGlob = import.meta.glob('../extensions/**/*.json', { eager: true }) as Record<string, any>

      // Find the correct locale file for this extension and language
      let localeFile = null
      for (const path in extensionGlob) {
        if (path.includes(`/${extensionName}_`) &&
            path.includes(`/locales/${languageCode}.json`)) {
          localeFile = path
          break
        }
      }

      if (localeFile && localeFile in extensionGlob) {
        const translationData = extensionGlob[localeFile].default

        // Merge extension translations directly into the root namespace
        const currentTranslations = this.translations.get(languageCode) || {}
        const merged = this.deepMerge(currentTranslations, translationData)
        this.translations.set(languageCode, merged)
        this.version.value++
      }
    } catch (error) {
      console.warn(`Failed to load translations for extension ${extensionName} in ${languageCode}:`, error)
    }
  }

  // Add custom translations to a language (public method for components)
  addTranslations(languageCode: string, translations: Translations) {
    const currentTranslations = this.translations.get(languageCode) || {}
    const merged = this.deepMerge(currentTranslations, translations)
    this.translations.set(languageCode, merged)
    this.version.value++
  }

  // Load extension translations for all enabled extensions
  async loadExtensionTranslationsForEnabledExtensions() {
    try {
      // Get enabled extensions from API
      const { default: http } = await import('@/utils/dynamic-http')
      const response = await http.get('/api/extensions/public')
      const extensions = response.data?.items || []

      const currentLanguage = this.getCurrentLanguage()

      // Load translations for each enabled extension
      for (const ext of extensions) {
        if (ext.name) {
          await this.loadExtensionTranslationsForExtension(ext.name, currentLanguage)
        }
      }
    } catch (error) {
      console.warn('Failed to load extension translations for enabled extensions:', error)
    }
  }


  // Set current language
  async setLanguage(languageCode: string) {
    if (languageCode === this.currentLanguage.value) return

    localStorage.setItem('preferredLanguage', languageCode)
    this.currentLanguage.value = languageCode

    await this.loadTranslations(languageCode)

    // Load extension translations for enabled extensions in the new language
    await this.loadExtensionTranslationsForEnabledExtensions()

    // Save to backend (async)
    try {
      await http.post('/api/user/language', { language: languageCode })
    } catch (error) {
      // Ignore backend errors
    }

    window.dispatchEvent(new CustomEvent('language-changed', { detail: { language: languageCode } }))
  }

  // Get current language
  getCurrentLanguage() {
    return this.currentLanguage.value
  }

  // Get version for reactivity
  getVersion() {
    return this.version.value
  }

  // Get available languages
  getAvailableLanguages() {
    const available = Array.from(this.translations.keys())
    // Always include English
    if (!available.includes('en')) available.push('en')
    return available
  }

  // Translate a key
  t(key: string, defaultValue?: string, params?: Record<string, string>): string {
    const translations = this.translations.get(this.currentLanguage.value) || {}

    // Try nested access (both EN and BG use nested structure)
    let translation = this.getNestedValue(translations, key)

    if (!translation) {
      // Try English as fallback if current language is not English
      if (this.currentLanguage.value !== 'en') {
        const englishTranslations = this.translations.get('en') || {}
        translation = this.getNestedValue(englishTranslations, key)
      }

      if (!translation) {
        return defaultValue || key
      }
    }

    // Replace parameters
    if (params) {
      for (const [param, value] of Object.entries(params)) {
        translation = translation.replace(`{${param}}`, value)
      }
    }

    return translation
  }

  // Get nested value using dot notation
  private getNestedValue(obj: any, key: string): string | null {
    const keys = key.split('.')
    let current = obj

    for (const k of keys) {
      if (current && typeof current === 'object' && k in current) {
        current = current[k]
      } else {
        return null
      }
    }

    return typeof current === 'string' ? current : null
  }

  // Check if language is available
  isLanguageAvailable(languageCode: string): boolean {
    return this.translations.has(languageCode)
  }
}

// Create global instance
export const i18n = new FrontendI18n()

// Helper function for components
export const useI18n = () => {
  const currentLanguage = computed(() => i18n.getCurrentLanguage())
  const availableLanguages = computed(() => i18n.getAvailableLanguages())
  const version = computed(() => i18n.getVersion())

  const t = (key: string, defaultValue?: string, params?: Record<string, string>) => {
    const _ = currentLanguage.value
    const __ = version.value
    return i18n.t(key, defaultValue, params)
  }

  const setLanguage = (language: string) => i18n.setLanguage(language)

  return {
    t,
    setLanguage,
    currentLanguage,
    availableLanguages
  }
}

// Initialize i18n
export const initializeI18n = async () => {
  const savedLanguage = localStorage.getItem('preferredLanguage')

  if (savedLanguage && i18n.isLanguageAvailable(savedLanguage)) {
    await i18n.setLanguage(savedLanguage)
  } else {
    // For new users, don't set localStorage yet, let loadDefaults set the default
    await i18n.setLanguage('en')
  }

  // Load extension translations for enabled extensions
  await i18n.loadExtensionTranslationsForEnabledExtensions()
}

File: backend/utils/i18n.py
"""
Internationalization (i18n) utilities
"""

import json
from pathlib import Path
from typing import Dict, Any, Optional
from sqlalchemy.orm import Session
from backend.database import get_db
from backend.db.language_pack import LanguagePack

class I18nManager:
    """Manages translations for different languages"""

    def __init__(self):
        self.translations: Dict[str, Dict[str, Any]] = {}
        self.current_language = "bg"  # Default to Bulgarian
        self._default_loaded = False

    def load_language_pack_from_db(self, language_code: str):
        """Load translations for a language from database"""
        try:
            db = next(get_db())
            language_pack = db.query(LanguagePack).filter(
                LanguagePack.code == language_code,
                LanguagePack.is_active == True
            ).first()
            
            if not language_pack:
                print(f"Language pack {language_code} not found in database")
                return False
            
            self.translations[language_code] = {}
            
            # Start with English as base if not exists
            if "en" not in self.translations:
                self.translations["en"] = self._get_default_english_translations()
            
            # Load frontend translations
            if language_pack.frontend_translations:
                self._merge_dict(self.translations[language_code], language_pack.frontend_translations)
            
            # Load backend translations
            if language_pack.backend_translations:
                self._merge_dict(self.translations[language_code], language_pack.backend_translations)
            
            # Load extensions translations
            if language_pack.extensions_translations:
                self._merge_dict(self.translations[language_code], language_pack.extensions_translations)
            
            print(f"Loaded {language_code} translations from database: {len(self.translations[language_code])} keys")
            db.close()
            return True
            
        except Exception as e:
            print(f"Error loading language pack {language_code} from database: {e}")
            return False

    def load_language_pack(self, language_code: str, translations_dir: Path):
        """Load translations for a language from a directory"""
        if not translations_dir.exists():
            print(f"Translations directory not found: {translations_dir}")
            return

        self.translations[language_code] = {}

        # Load all json files in the translations directory
        for json_file in translations_dir.glob("*.json"):
            try:
                with open(json_file, 'r', encoding='utf-8') as f:
                    data = json.load(f)
                    # Merge into the language dict
                    self._merge_dict(self.translations[language_code], data)
            except Exception as e:
                print(f"Error loading translation file {json_file}: {e}")

        print(f"Loaded {language_code} translations: {len(self.translations[language_code])} keys")

    def _get_default_english_translations(self) -> Dict[str, Any]:
        """Get default English translations as fallback"""
        return {
            "api": {
                "errors": {
                    "unauthorized": "Unauthorized access",
                    "forbidden": "Access forbidden",
                    "not_found": "Resource not found",
                    "validation_error": "Validation error",
                    "internal_error": "Internal server error",
                    "bad_request": "Bad request",
                    "server_error": "Server error"
                },
                "messages": {
                    "success": "Operation completed successfully",
                    "created": "Resource created successfully",
                    "updated": "Resource updated successfully",
                    "deleted": "Resource deleted successfully",
                    "saved": "Data saved",
                    "loaded": "Data loaded",
                    "exported": "Data exported",
                    "imported": "Data imported"
                }
            },
            "extensions": {
                "messages": {
                    "installed": "Extension installed successfully",
                    "uninstalled": "Extension uninstalled successfully",
                    "enabled": "Extension enabled",
                    "disabled": "Extension disabled",
                    "updated": "Extension updated",
                    "uploaded": "Extension uploaded",
                    "deleted": "Extension deleted"
                }
            },
            "language": {
                "messages": {
                    "detected": "Language automatically detected",
                    "switched": "Language switched to {language}",
                    "loaded": "Language files loaded",
                    "cached": "Translations cached"
                }
            }
        }

    def _merge_dict(self, target: Dict[str, Any], source: Dict[str, Any]):
        """Recursively merge source dict into target"""
        for key, value in source.items():
            if isinstance(value, dict) and key in target and isinstance(target[key], dict):
                self._merge_dict(target[key], value)
            else:
                target[key] = value

    def set_language(self, language_code: str):
        """Set the current language"""
        if language_code == "en":
            self.current_language = language_code
            if not self._default_loaded:
                self.translations["en"] = self._get_default_english_translations()
                self._default_loaded = True
        elif language_code in self.translations:
            self.current_language = language_code
        else:
            # Try to load from database
            if self.load_language_pack_from_db(language_code):
                self.current_language = language_code
            else:
                print(f"Language {language_code} not available, falling back to en")
                self.current_language = "en"
                if not self._default_loaded:
                    self.translations["en"] = self._get_default_english_translations()
                    self._default_loaded = True

    def get(self, key: str, default: str = None, language: str = None) -> str:
        """Get a translated string"""
        lang = language or self.current_language

        # Try the specified language
        if lang in self.translations:
            value = self._get_nested_value(self.translations[lang], key)
            if value is not None:
                return value

        # Fallback to English (default)
        if lang != "en" and "en" in self.translations:
            value = self._get_nested_value(self.translations["en"], key)
            if value is not None:
                return value

        # Return default or the key itself
        return default or key

    def _get_nested_value(self, data: Dict[str, Any], key: str) -> Optional[str]:
        """Get a nested value from dict using dot notation"""
        keys = key.split('.')
        current = data

        for k in keys:
            if isinstance(current, dict) and k in current:
                current = current[k]
            else:
                return None

        return current if isinstance(current, str) else None

    def get_current_language(self) -> str:
        """Get the current language code"""
        return self.current_language

    def get_available_languages(self) -> list:
        """Get list of available language codes"""
        try:
            db = next(get_db())
            language_packs = db.query(LanguagePack.code).filter(
                LanguagePack.is_active == True
            ).all()
            db.close()

            codes = [pack.code for pack in language_packs]
            # Always include English
            if "en" not in codes:
                codes.append("en")
            return codes
        except Exception as e:
            print(f"Error getting available languages: {e}")
            # Fallback to loaded translations
            return list(self.translations.keys()) + ["en"]

# Global instance
i18n_manager = I18nManager()

# Convenience functions
def _(key: str, default: str = None, language: str = None) -> str:
    """Translate a key"""
    return i18n_manager.get(key, default, language)

def set_language(language_code: str):
    """Set the current language"""
    i18n_manager.set_language(language_code)

def get_current_language() -> str:
    """Get the current language"""
    return i18n_manager.get_current_language()